### DISCUSSION TITLE
Extend CA Trust capabilities

### DISCUSSION THREAD

**Discussion by boyobejamin** [mitmproxy/mitmproxy#5743] [Q&A]
#### Problem Description

Users debugging or inspecting websites with self signed certificates (those not publically trusted), cannot provide a method for `mitmproxy` tools to trust their CAs and certificates. The only option is to us `-k` to opt-out of certificate verification.

#### Proposal

Allow for users to specify their own trusted CAs for `mitmproxy` and tools to trust upstream. This could be leveraged a few ways to include users providing a root ca pem file, using the system CA trust provided on Ubuntu/RHEL, or a directory including certificates to be trusted (e.g. `/etc/ssl/certs` or `/etc/pki/tls/anchors`). 

This would allow for users who internally sign certificates to validate the certificate chain as part of their debugging processes when using these tools.

#### Alternatives

It would be nice to leverage the systems trust, it would be a fair assumption that the system already trusts internal certificates the user is attempting to debug.

#### Additional context
Add any other context or screenshots about the proposal here.

**Comment by mhils** [comment]
Are you aware of https://docs.mitmproxy.org/dev/concepts-options/#ssl_verify_upstream_trusted_ca and https://docs.mitmproxy.org/dev/concepts-options/#ssl_verify_upstream_trusted_confdir?

Trust store is tricky, but I'd be happy to accept a proof-of-concept patch.

**Comment by boyobejamin** [comment]
Hey there!

Yes!

`ssl_verify_upstream_trusted_ca` expects that I also have the private key associated with my CA. I'd like to propose this feature from the perspective of operating in a mature organization where an engineer doesn't have access to the organizations cert/pem.

`ssl_verify_upstream_trusted_confdir` doesn't work as I am expecting. It could be misunderstanding how it's intended to operate or a bug. I'm trying to pass it either a directory containing the organizations individual root and intermediate ca files OR `/etc/ssl/certs/` on Ubuntu. I'd prefer passing the latter as it already has the c hash'ed files as requested. Only works as a yaml config option, can't pass it as a command line arg.

**Comment by mhils** [comment]
> `ssl_verify_upstream_trusted_ca` expects that I also have the private key associated with my CA.

I don't think it does...

> Only works as a yaml config option, can't pass it as a command line arg.

`--set config_option=foo` works for all options (although I agree we should eventually do a better job here).

**Comment by boyobejamin** [comment]
I've been using `mitmproxy` for an hour. I trust that I don't know what I doing and know enough to be dangerous 😂 

Tracking with set, I did try it and no dice. Or at least, I didn't get the result I desired.
